### ISSUE TITLE
Fix S2328 FP: GetHashCode on mutable struct is safe

### ISSUE THREAD

**Issue by jilles-sg** [SonarSource/sonar-dotnet#8756]
### Description

Rule S2328 "GetHashCode" should not reference mutable fields complains about using mutable fields in the `GetHashCode` method because mutating the field may cause the object to become misfiled in a hash table. However, this scenario can only happen for classes, not structs. When a struct instance is mutated, this does not affect the hash table's copy (other than the obscure case of the `Equals` or `GetHashCode` methods themselves changing the hash code).

The designers of C# apparently also think it is OK to have mutable structs with `GetHashCode`, for example `ValueTuple<T1, T2>` and `record struct`.

### Repro steps

```c#
struct S: IEquatable<S> {
    public int A;
    public int B;
    public override bool Equals(object? obj) => obj is S other && Equals(other);
    public bool Equals(S other) => A == other.A && B == other.B;
    public override int GetHashCode() => A;
}
```
### Expected behavior

No warning.

### Actual behavior

Rule S2328 complains.

### Known workarounds

* Make the fields readonly, which requires additional lines of code for little benefit.
* Have the C# compiler generate `GetHashCode` by making the type a `record struct`.
* Use a tuple (`ValueTuple`) instead of a custom struct.

### Related information

* C#/VB.NET Plugins version: csharp version 9.19.0.84025
* Visual Studio version: 2022
* MSBuild / dotnet version: .NET 6.0
* SonarScanner for .NET version (if used): SonarScanner for MSBuild 5.14 (.NET Core) with SonarCloud
* Operating System: Windows 10.0.14393

**Comment by gregory-paidis-sonarsource** [comment]
Hey there!
That's a very interesting point.
I had an internal  discussion with my team, and we agree with your reasoning.
I accept the suggestion and a fix is on the way :)

**Comment by gregory-paidis-sonarsource** [comment]
The issue has been fixed!
You will be able to find the new rule behavior in the next SonarQube, SonarCloud and SonarLint release.
